nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDBRecord.java

```diff
@@ -0,0 +1,303 @@
+                            .valueOf(LocalDateTime.parse((String) values[0], format))
+                            .getTime();
+                } else {
+                    timestamp = (Long) values[0];
+                }
+
+                boolean isFulled = false;
+
+                for (Map.Entry<String, Tablet> entry : tablets.entrySet()) {
+                    String device = entry.getKey();
+                    Tablet tablet = entry.getValue();
+                    int rowIndex = tablet.rowSize++;
+
+                    tablet.addTimestamp(rowIndex, timestamp);
+                    List<MeasurementSchema> measurements = tablet.getSchemas();
+                    for (MeasurementSchema measurement : measurements) {
+                        String tsName =
+                                new StringBuilder()
+                                        .append(device)
+                                        .append(".")
+                                        .append(measurement.getMeasurementId())
+                                        .toString();
+                        int valueIndex = fieldNames.indexOf(tsName) + 1;
+                        Object value;
+                        TSDataType type = measurement.getType();
+                        if (values[valueIndex] != null) {
+                            try {
+                                value = convertType(values[valueIndex], type);
+                            } catch (Exception e) {
+                                getLogger().error("The value {{}} can't be converted to the type {{}}", values[valueIndex], type);
+                                processSession.transfer(flowFile, REL_FAILURE);
+                                return;
+                            }
+                        } else
+                            value = null;
+                        tablet.addValue(measurement.getMeasurementId(), rowIndex, value);
+                    }
+                    isFulled = tablet.rowSize == tablet.getMaxRowNumber();
+                }
+                if (isFulled) {
+                    if (aligned) {
+                        session.get().insertAlignedTablets(tablets);
+                    } else {
+                        session.get().insertTablets(tablets);
+                    }
+                    tablets.values().forEach(tablet -> tablet.reset());
+                }
+            }
+
+            AtomicBoolean hasRest = new AtomicBoolean(false);
+            tablets.forEach(
+                    (device, tablet) -> {
+                        if (hasRest.get() == false && tablet.rowSize != 0) {
+                            hasRest.set(true);
+                        }
+                    });
+            if (hasRest.get()) {
+                if (aligned) {
+                    session.get().insertAlignedTablets(tablets);
+                } else {
+                    session.get().insertTablets(tablets);
+                }
+            }
```

Comment:
This `try` block is large and has many error conditions. It would be easier to follow and maintain if the various checks could be moved to separate methods, which would throw exceptions. That would allow any problems to be caught by the outer catch, and result in the FlowFile being routed to failure.